src/little_dict.jl

```diff
@@ -0,0 +1,245 @@
+    if key ∈ dd.keys
+        return key
+    else
+        return default
+    end
+end
+
+struct NotFoundSentinel end  # Struct to mark not not found
+function Base.get(dd::LittleDict, key, default)
+    @assert length(dd.keys) == length(dd.vals)
+    for ii in 1:length(dd.keys)
+        cand = @inbounds dd.keys[ii]
+        isequal(cand, key) && return @inbounds(dd.vals[ii])
+    end
+    return default
+end
+function get(default::Base.Callable, dd::LittleDict, key)
+    got = get(dd, key, NotFoundSentinel())
+    if got isa NotFoundSentinel  # not found
+        return default()
+    else
+        return got
+    end
+end
+
+function Base.iterate(dd::LittleDict, ii=1)
+    ii > length(dd.keys) && return nothing
+    return (dd.keys[ii] => dd.vals[ii], ii+1)
+end
+
+function merge(d1::LittleDict, d2::AbstractDict)
+    return merge((x,y)->y, d1, d2)
+end
+
+function merge(
+    combine::Function,
+    d::LittleDict,
+    others::AbstractDict...
+    )
+    K,V = _merge_kvtypes(d, others...)
+    dc = LittleDict{K,V}(d)
+    for d2 in others
+        for (k2,v2) in d2
+            got = get(dc, k2, NotFoundSentinel())
+            if got isa NotFoundSentinel
+                add_new!(dc, k2, v2)
+            else
+                # GOLDPLATE: ideally we would avoid iterating this twice
+                # once for get and once for setindex!
+                dc[k2]=combine(got, v2)
+            end
+        end
+    end
+    return dc
+end
+
+
+Base.empty(dd::LittleDict{K,V}) where {K,V} = LittleDict{K,V}()
+
+######## Methods that all mutable AbstractDict's should implement
+
+function Base.sizehint!(dd::LittleDict, sz)
+    sizehint!(dd.keys, sz)
+    sizehint!(dd.vals,sz)
+    return dd
+end
+
+function add_new!(dd::LittleDict, key, value)
+    # Not found, add to the end
+    push!(dd.keys, key)
+    try
+        push!(dd.vals, value)
+    catch
+        # if we sucessfully added  a key, but failed to add a value
+        # then we need to remove the key so the little dict remains with constistant state
+        pop!(dd.keys)
+        rethrow()
+    end
+    return dd
+end
+
+
+# This is the safe case that can't error on the adding of values.
+# So no catching exceptions then rollback of adding keys is needed
+# Optimising this case gives a 30% speedup on setindex! when element is new
+# On a vaguely realistic microbenchmark
+function add_new!(dd::LittleDict{<:Any,V,<:Vector,<:Vector{V}}, key, value::V) where {V}
+    # Not found, add to the end
+    push!(dd.keys, key)
+    push!(dd.vals, value)
+    return dd
+end
+
+function Base.setindex!(dd::LittleDict, value, key)
+    # Assertion below commented out as by standards of carefully optimised
+    # setindex! it has huge code (26%), this does mean that if someone has messed
+    # with the fields of the LittleDict directly, then the @inbounds could be invalid
+    #@assert length(dd.keys) == length(dd.vals)
+    @simd for ii in 1:length(dd.keys)
+        cand = @inbounds dd.keys[ii]
```

Comment:
Just curious, is this `@inbounds` necessary?  I would have thought that the compiler would notice that we're iterating over the indexes of `dd.keys` and realized that this was inbounds.

(I want to check this myself to understand it better, but I have to get back to real work in a minute...)

**Edit**: Okay, I checked, and yes it is necessary.  But for my (very simple) test, the `@simd` didn't change anything.  Maybe you could include your microbenchmark somewhere?

```julia
using BenchmarkTools

using OrderedCollections

function test(d)
    for i=1:10000
        for c = 'a':'z'
            d[c] = rand(Int)
        end
    end
    return d
end

const d = LittleDict{Char, Int}()
@btime test($d)
```

Reply:
Also, if I run the script above from the command line, I get some interesting timings using `eachindex(dd.keys)` instead of `1:length(dd.keys)`.
About half the time, it's slightly faster (2.6 ms with `eachindex` vs 2.8 ms with `length` on my machine), and about half the time it's significantly slower (~3.3-3.8 ms with `eachindex`).
Timings with `length` are more consistent (2.8-2.9ms).

 :-/

This is with stock Julia 1.1 on a MacBook Pro.

Reply:
Yes, should remove `@simd` I thought i did during my benchmarking last night, but I must have added it back by mistake.